### PR TITLE
Gridnet dev

### PR DESCRIPTION
**Pull request by boxofpasta**

### CODE REVIEW

File: common/models.py [hunk 1]
```diff
@@ -1,16 +1,20 @@
 import tensorflow as tf
+from utils.misc import print_tensor_shape
 
+_default = object()
 
 class ConvNetwork:
     def __init__(self, layer_specs=None,
                  activation_fn=tf.nn.leaky_relu,
+                 last_activation_fn=_default,
```
Comment (NeedsMorePie) [inline common/models.py]: I've been dealing with the last activation function by appending an extra "output layer" after calling _get_conv_tower. If we update ConvNetwork to have a last_activation_fn, we need to update the other networks to use it.

File: gridnet/connections/connections.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        :return: Tensor. Feature map of shape [batch_size, H, W, num_features].
+        """
+        with tf.variable_scope(self.name, reuse=reuse_variables):
+
+            # Pass through resolution preserving convolutions, with skip connection at the end.
+            if self.activation_fn is not None:
+                previous_output = self.activation_fn(features)
+
+            previous_output, layer_outputs = self._get_conv_tower(previous_output)
+            final_output = features + previous_output
```
Comment (NeedsMorePie) [inline gridnet/connections/connections.py]: Is this supposed to be an addition or a concatenation? I can't tell from the context-aware paper.

File: gridnet/gridnet.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                    if i == 0 and j == 0:
+                        continue
+
+                    top_output, left_output = 0, 0
+                    if i > 0:
+                        top_output = self.process_downwards(grid_outputs[i-1][j], i, j)
+                    if j > 0:
+                        left_output = self.process_rightwards(grid_outputs[i][j-1], i, j)
+
+                    grid_outputs[i][j] = top_output + left_output
```
Comment (NeedsMorePie) [inline gridnet/gridnet.py]: Same question again, are these supposed to be additions?

File: gridnet/gridnet.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+                    grid_outputs[i][j] = bottom_output + left_output
+
+            # Final lateral connection.
+            previous_output = grid_outputs[0][self.width-1]
+            final_output = self.process_rightwards(previous_output, 0, self.width)
+            return final_output, grid_outputs
+
+    # Helper functions.
+    def process_rightwards(self, input, i, j):
```
Comment (NeedsMorePie) [inline gridnet/gridnet.py]: For private functions, use _process_rightwards with an extra _ at the front.

File: common/models.py (hunk 1 above)
Comment (boxofpasta) [reply common/models.py]: I think that it's common enough of a use case that we should have it as an option. It's not meant to be necessary and should be compatible with what you already have (i.e should just default to the value of activation_fn).

File: gridnet/connections/connections.py (hunk 2 above)
Comment (boxofpasta) [reply gridnet/connections/connections.py]: These use the same ops as the grid intersections outputs -- they are additions.

File: gridnet/gridnet.py (hunk 3 above)
Comment (boxofpasta) [reply gridnet/gridnet.py]: Quoting the GridNet [paper](https://arxiv.org/pdf/1707.07958.pdf)(bottom of page 5):

"Apart from border blocks, each feature map Xi, j in the grid is the result of two different
computations: one horizontal residual computation processing data from Xi, j−1 and
one vertical computation processing data from Xi−1, j or Xi+1, j depending if the column is a
subsampling or upsampling one. Several choices can be taken here, including concatenating
features, summing, or learning a fusion. We opted for summing, a choice which keeps
model capacity low and blends well with the residual nature of the grid streams."

File: gridnet/gridnet.py (hunk 4 above)
Comment (boxofpasta) [reply gridnet/gridnet.py]: Will do.

**Pull request merged**